rancherdesktop/instanceInfo.go

```diff
@@ -0,0 +1,81 @@
+package rancherdesktop
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+)
+
+var validPlatforms map[string]struct{} = map[string]struct{}{
+	"win32":  {},
+	"darwin": {},
+	"linux":  {},
+}
+
+var validArchs map[string]struct{} = map[string]struct{}{
+	"x64":   {},
+	"arm64": {},
+}
+
+type CheckUpgradeRequest struct {
+	AppVersion string            `json:"appVersion"`
+	ExtraInfo  map[string]string `json:"extraInfo"`
+}
+
+// InstanceInfo contains all the info we need about an instance of Rancher Desktop.
+// Note that older versions of Rancher Desktop cannot be represented by
+// an InstanceInfo struct, since they do not send certain required fields.
+type InstanceInfo struct {
+	AppVersion      *semver.Version
+	Platform        string
+	Arch            string
+	PlatformVersion *semver.Version
```

Comment:
They are pointers because `*semver.Version` is what `semver.NewVersion` returns. It's also what is used in places like `semver.Constraints.Check`. I guess these variables don't *need* to be pointers, but it is more convenient if they are, given that Go will automatically dereference pointers, but it won't automatically prepend a `&` to value variables.

The only place an `InstanceInfo` is used is when coming up with a response to a request from Rancher Desktop. In this case, a `*CheckUpgradeResponse` is parsed into an `InstanceInfo` via `NewInstanceInfo()`, which validates the input in such a way that neither `InstanceInfo.AppVersion` or `InstanceInfo.PlatformVersion` can be `nil`.

We can change this if you want, but it doesn't seem necessary to me.